### ISSUE TITLE
Improve addSource(File source) to accept package-info.java files as well.

### ISSUE THREAD

**Issue by lennartj** [paul-hammant/qdox#5]
When using QDox to extract information from a ```package-info.java``` file, we currently have to use the ```addSource(URL)``` method, since the ```addSource(File)``` simply ignores package-info.java files. This is not documented, as far as I can see. The code snippet below is from SourceLibrary.java, and it has opinions about the package-info.java when added as a File - but parses it happily when added as an URL:

```java
    /**
     * Add a {@link URL} containing java code to this library
     * 
     * @param url a {@link URL} which should contain java code
     * @return The constructed {@link JavaSource} object of this url
     * @throws ParseException if this content couldn't be parsed to a JavaModel
     */
    public JavaSource addSource( URL url )
        throws ParseException, IOException
    {
        return addSource( new InputStreamReader( url.openStream(), encoding), url );
    }

    /**
     * Add a {@link File} containing java code to this library
     * 
     * @param file a {@link File} which should contain java code
     * @return The constructed {@link JavaSource} object of this file
     * @throws ParseException
     * @throws IOException
     */
    public JavaSource addSource( File file )
        throws ParseException, IOException
    {
        JavaSource result = null;
        if ( !"package-info.java".equals( file.getName() ) ) 
        {
            result = parse( new FileInputStream( file ), file.toURI().toURL() );
            // if an error is handled by the errorHandler the result will be null
            if( result != null )
            {
                if( getJavaPackage( result.getPackageName() ) == null )
                {
                    File packageInfo = new File(file.getParentFile(), "package-info.java");
                    if( packageInfo.exists() )
                    {
                        JavaPackage pckg = parse( new FileInputStream( packageInfo ), packageInfo.toURI().toURL() ).getPackage();
                        context.add( pckg );
                    }
                }
                registerJavaSource(result);
            }
        }
    	return result;
    }
```

This leads to somewhat clunky code on the client side:

```java
if(sourceCodeFile.getName().trim().equalsIgnoreCase("package-info.java")) {

                // For some reason, QDox requires the package-info.java to
                // be added as a URL instead of a File.
                builder.addSource(sourceCodeFile.toURI().toURL());
                ....
```

**Comment by rfscholte** [comment]
Good catch!

**Comment by rfscholte** [comment]
Just trying to remember why I wrote it like this. IIRC package-info is not considered a JavaSource file, since it can't contain classes, etc. It contains metadata for all classes within this same package. So if there is a java file in this package, this code will search for the package-info and bind it when available.
What I'm missing is what you want: access to the JavaPackage, even if there are no Java files?

**Comment by lennartj** [comment]
Well, I'll take one specific example where I ran into this particular behavior. I'm using QDox internally within the jaxb2-maven-plugin to extract two things:

1. JavaDoc to inject into generated XSDs (JavaDoc can be found on packages as well).
2. Which sources should be included in the SchemaGen or XJC compilation. In this sense, the plugin can filter sources by examining which ones are valid for JAXB compilation.

In both of these cases, QDox may need to munch package-info.java, since some JAXB annotations may be found there.

**Comment by lennartj** [comment]
A reference to the actual code is in the 

https://github.com/mojohaus/jaxb2-maven-plugin/blob/master/src/main/java/org/codehaus/mojo/jaxb2/schemageneration/AbstractXsdGeneratorMojo.java

in the method

````java
private List<String> getSchemaGeneratorSourceFiles(final List<URL> sources)
            throws IOException, MojoExecutionException { ... } 
````

**Comment by rfscholte** [comment]
I don't think you should do this on a per-file or per-URL based action. Why not use ProjectBuilder.addSourceTree(), followed by ProjectBuilder.getSources() and ProjectBuilder.getPackages() ? The latter will give you all that package-info.java details.

**Comment by lennartj** [comment]
I don't think that works, since the plugin has filters to exclude files/directories matching supplied regular expressions. If I do addSourceTree(), wouldn't I include all files below that particular directory (i.e. disregarding any provided patterns for file or directory exclusions)?

**Comment by rfscholte** [comment]
Did you try the `ProjectBuilder.getPackages()` after you're done?
I hope we can work on a directory-based solution. Otherwise we need to add something like `public JavaPackage addPackage(File);`

**Comment by lennartj** [comment]
The current solution I have works well.
However, I had to take a look in the QDox sources to understand why certain package annotations were not included in the standard compilation units.

If this is indeed the intended behavior of QDox, I suggest it should be documented in a clear way, to avoid people scratching their heads wondering why some things didn't quite work as they expected.

**Comment by rfscholte** [comment]
URL and File based should work the same (that's why I started with "Good catch"), so I expect to improve the URL implementation as well. Then you can't use this method the way you want it.
